### PR TITLE
Support nesting icu params

### PR DESCRIPTION
**Pull request by yuval-zamir**
What are the changes?
(1) Solves an issue for nesting params in ICU format: 
For example, for the following ICU text `Hey, {numPersons, plural, =0 {No one here.} one {{name}. You are the only person here.} other {{name} and # other persons are here.}}`. the current logic is mistakenly not recognizing the `name` param since its inside the `numPersons` param. This will be resulted as `Record<'numPersons'>` instead of  `Record<'numPersons' | 'name'>`

~~(2) Add types to the params. Using the param type we can tell the consumer if a parameter is number/string/etc. So instead of having `Record<'name' | 'age'>` we'll have now `{ name?: string; age?: number }`~~

** Please note that the changes are only for the single curly braces case because it's the only way to support ICU.

#pr-cr@titk,guyw

### CODE REVIEW

**Comment by varzager** [comment]
@yuval-zamir looks great!, and also is a great fix! 

- but since you also are making it typed this has to be a `major`, otherwise it will break all projects using it.
   If we are going to make a major, so maybe we should use the opportunity to make more changes. such as get rid of double curly braces. change defaults, and maybe more. 
  - obviously we should also make a migration guide
  -  you can leave it without any types and then we will do it some other time.
- Please add tests for the nested case. not only add to the json, but also have `it` testing it, and it's types.

File: tests/sources/icu.json [hunk 1]
```diff
@@ -2,7 +2,8 @@
   "common": {
     "people": {
       "message": "Hey, {numPersons, plural, =0 {no one} =1 {one person} other {# persons}}",
-      "messageComplex": "Hey {name}, There are {numPersons, plural, =0 {no one} =1 {one person} other {# persons}} that want to change the {productsAmount, plural, =1 {price of 1 product} other {prices of # products}}"
+      "messageComplex": "Hey {name}, There are {numPersons, plural, =0 {no one} =1 {one person} other {# persons}} that want to change the {productsAmount, plural, =1 {price of 1 product} other {prices of # products}}",
+      "messageNestingParams": "Hey, {numPersons, plural, =0 {No one here.} one {{name}. You are the only person here.} other {{name} and # other persons are here.}}"
```
Comment (varzager) [inline tests/sources/icu.json:6]: please add a `it` that tests this, and it's types

**Review by varzager — CHANGES_REQUESTED** [review]

**Comment by yuval-zamir** [comment]
> @yuval-zamir looks great!, and also is a great fix!
> 
> * but since you also are making it typed this has to be a `major`, otherwise it will break all projects using it.
>   If we are going to make a major, so maybe we should use the opportunity to make more changes. such as get rid of double curly braces. change defaults, and maybe more.
>   
>   * obviously we should also make a migration guide
>   * you can leave it without any types and then we will do it some other time.
> * Please add tests for the nested case. not only add to the json, but also have `it` testing it, and it's types.

Argee that we have the types changes as a Major so I maybe split this PR ti the nested tyeps bug fix and the types as a separated PR.

**Comment by yuval-zamir** [comment]
#skipreview

**Review by yuval-zamir — COMMENTED** [review]
Changes have been made. Will create new PR for major version

File: src/icuParams.ts [hunk 2]
```diff
@@ -0,0 +1,92 @@
...
+  let stack = parsedArray.slice();
+  const params: Param[] = [];
+  const used = new Set();
+  while (stack.length) {
+    const element = stack.pop();
+    if (isStringOnly(element as Format)) continue;
+    if (hasHashtagOnly(element)) continue;
+
+    const [name, format] = element!;
+    const subMessages = getSubMessages(element, format as Format);
```
Comment (varzager) [inline src/icuParams.ts:72]: move after `if(!used.has(name))`

File: tests/sources/icu-nested.json [hunk 3]
```diff
@@ -0,0 +1,8 @@
+{
+  "common": {
+    "people": {
+      "message": "Hey, {numPersons, plural, =0 {no one} =1 {one person} other {# persons}}",
+      "messageNestedParams": "Hey, {numPersons, plural, =0 {No one here.} one {{name}. You are the only person here.} other {{name} and # other persons are here.}}"
+    }
```
Comment (varzager) [inline tests/sources/icu-nested.json:6]: add cases for all formats - 
  | 'selectordinal'
  | 'date'
  | 'time'
  | 'select'
  | 'number'

**Review by varzager — CHANGES_REQUESTED** [review]

File: tests/sources/icu-nested.json (hunk 3 above)
Comment (yuval-zamir) [reply tests/sources/icu-nested.json:6]: Fixed

File: src/icuParams.ts (hunk 2 above)
Comment (yuval-zamir) [reply src/icuParams.ts]: Fixed

**Review by varzager — APPROVED** [review]

**Pull request merged**